### PR TITLE
add Flatten() for iterating ExpandResponses

### PR DESCRIPTION
**Pull request by jzelinskie**

### CODE REVIEW

File: flatten.go [hunk 1]
```diff
@@ -0,0 +1,159 @@
...
+		switch typed.IntermediateNode.Operation {
+		case api.SetOperationUserset_UNION:
+			return flattenUnion(typed.IntermediateNode.ChildNodes)
+		case api.SetOperationUserset_INTERSECTION:
+			return flattenIntersection(typed.IntermediateNode.ChildNodes)
+		case api.SetOperationUserset_EXCLUSION:
+			return flattenExclusion(typed.IntermediateNode.ChildNodes)
+		}
+	case *api.RelationTupleTreeNode_LeafNode:
+		var toReturn userSet = make(map[string]struct{})
```
Comment (josephschorr) [inline flatten.go:25]: `toReturn := newUserSet()` and then add a method

File: flatten.go [hunk 2]
```diff
@@ -0,0 +1,159 @@
...
+	return setResult(api.SetOperationUserset_INTERSECTION, start, children)
+}
+
+func exclusion(start *api.ObjectAndRelation, children ...*api.RelationTupleTreeNode) *api.RelationTupleTreeNode {
+	return setResult(api.SetOperationUserset_EXCLUSION, start, children)
+}
+
+type userSet map[string]struct{}
+
+func (us userSet) Add(users ...*api.User) {
```
Comment (josephschorr) [inline flatten.go:124]: Doc comments on anything exported

Comment (jzelinskie) [reply flatten.go:124]: It's on a non-exported type -- this isn't actually exported. I'll lowercase it.

File: flatten.go (hunk 1 above)
Comment (jzelinskie) [reply flatten.go:25]: do you think we really need this for something private?

File: flatten.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+
+func flatten(node *api.RelationTupleTreeNode) []*api.User {
+	switch typed := node.NodeType.(type) {
+	case *api.RelationTupleTreeNode_IntermediateNode:
+		switch typed.IntermediateNode.Operation {
+		case api.SetOperationUserset_UNION:
+			return flattenUnion(typed.IntermediateNode.ChildNodes)
+		case api.SetOperationUserset_INTERSECTION:
+			return flattenIntersection(typed.IntermediateNode.ChildNodes)
+		case api.SetOperationUserset_EXCLUSION:
```
Comment (josephschorr) [inline flatten.go:25]: Add a warning if we encounter an unknown type?

File: flatten.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+			return flattenExclusion(typed.IntermediateNode.ChildNodes)
+		}
+	case *api.RelationTupleTreeNode_LeafNode:
+		users := newUserSet()
```
Comment (josephschorr) [inline flatten.go:29]: Maybe have a `.extend` that takes in a slice?

Edit: or just pass `typed.LeafNode.Users...` to `add`?

File: flatten.go [hunk 5]
```diff
@@ -0,0 +1,167 @@
...
+	inOthers := newUserSet()
+	inOthers.add(flattenIntersection(children[1:])...)
+
+	maxChildren := len(firstChildChildren)
+	if len(inOthers) < maxChildren {
+		maxChildren = len(inOthers)
+	}
+
+	toReturn := make([]*api.User, 0, maxChildren)
+	for _, child := range firstChildChildren {
```
Comment (josephschorr) [inline flatten.go:62]: Perhaps add `intersect` and `subtract` to the user set itself?

File: flatten.go (hunk 3 above)
Comment (jzelinskie) [reply flatten.go:25]: I've changed the signature to return an error. When we figure out the desirable way to handle future gRPC API changes, we can use that error.

File: flatten.go (hunk 5 above)
Comment (jzelinskie) [reply flatten.go:60]: Those are both operations on slices right now to avoid allocations. I think we're best with what we've got.

**Review by josephschorr — APPROVED** [review]
LGTM

**Pull request merged**